scheduler/teachercourse_csp.py

```diff
@@ -249,7 +250,7 @@ def no_time_clash(val1, val2, course1=course_n):
         for d in WEEKDAYS:
             csp = CSP()
             courses = daily_courses[d]
-            add_nodes()
+            add_nodes(courses, rooms, rooms_chosen, full_prof_assignment, prof_info, csp)
```

Comment:
How was it working before when the arguments were not passed?
What changed that you have to pass the arguments now?

Reply:
Sorry didn't see your comment before I merged. It was working before because the function 'add_nodes' is nested in another function, where all these parameters are defined. To make it more readable and testable, I took it out of the other function and therefore needed to pass along the argument.

Reply:
If the function 'add_nodes' was only called inside the function it was nested in then why do we need it outside? I don't understand how does it improve readability? Can you explain me that?

Another thing you said to make it more testable. How was the 'add_nodes' function tested before?

Note: Please add a detailed explanation about what refactor was done and why it was needed. So that everyone can understand looking at the description.
Thanks

Reply:
If you look at the file "teachercourse_csp.py" in its initial state before any refactor, it only contains one function, "assigner." Inside this function, you will then have about ~10 helper functions defined inside it. Then in 2 of those helper functions, you would see even more helper functions defined within them. 

As for readability, I think extracting the helper functions out of "assigner" reduces code complexity. Since we are pushing for 100% unit test coverage, I think it would also help with testing by extracting the helper functions first out of the "assigner" so that James could write unit tests for these helper functions. There was no test in place for add_nodes.

I'll include explanation in my future PR. Thanks!

Reply:
Ok, sounds good!
Just out of curiosity, can't we create another class and define all the helper functions and then use it? Just something to think about maybe.